### PR TITLE
Make SynAxis produce an updating MoveStatus

### PR DESCRIPTION
**Pull request by callumforrester**
* Change `SynAxis` to make a `MoveStatus` rather than a `DeviceStatus`.
* Update this status periodically during motion (period is customisation with a constructor parameter, defaults to 6 times per move)
* Example use: progress bars with Bluesky

### CODE REVIEW

File: ophyd/sim.py [hunk 1]
```diff
@@ -418,39 +420,51 @@ def readback_func(x):
         super().__init__(name=name, parent=parent, labels=labels, kind=kind,
                          **kwargs)
         self.readback.name = self.name
+        self._events_per_move = events_per_move
+        self.egu = egu
+
+    def set(self, value: float) -> None:
```
Comment (tacaswell) [inline ophyd/sim.py:426]: This returns a Status object not None

File: ophyd/sim.py [hunk 2]
```diff
@@ -418,39 +420,51 @@ def readback_func(x):
         super().__init__(name=name, parent=parent, labels=labels, kind=kind,
                          **kwargs)
         self.readback.name = self.name
+        self._events_per_move = events_per_move
+        self.egu = egu
+
+    def set(self, value: float) -> StatusBase:
```
Comment (ZLLentz) [inline ophyd/sim.py:426]: Small detail: we can be more precise here
```suggestion
    def set(self, value: float) -> MoveStatus:
```

**Review by ZLLentz — APPROVED** [review]
I checked out this branch to give it a spin- works as advertised. This could be very useful.

We've accumulated merge conflicts since this PR was created.

**Comment by klauer** [comment]
Fixed the merge conflict

**Comment by callumforrester** [comment]
After some thought, I've changed the default value of `events_per_move` to 1 so that existing `SynAxis` objects behave exactly as before and deliberate intervention is needed to invoke the new behaviour (e.g. `SynAxis(name="x", events_per_move=6)`).

@tacaswell @ZLLentz Let me know if you have any thoughts on this

File: ophyd/sim.py (hunk 2 above)
Comment (callumforrester) [reply ophyd/sim.py]: Interestingly I thought this wouldn't typecheck but turns out it does: https://mypy-play.net/?mypy=latest&python=3.7&gist=9b1b36bb342d46dacc8296f77746ee8e

File: ophyd/sim.py [hunk 3]
```diff
@@ -422,6 +426,8 @@ def readback_func(x):
         super().__init__(name=name, parent=parent, labels=labels, kind=kind,
                          **kwargs)
         self.readback.name = self.name
+        if events_per_move < 1:
+            raise KeyError("At least 1 event per move is required")
```
Comment (klauer) [inline ophyd/sim.py:430]: `KeyError` seems like the wrong exception here; `ValueError` instead perhaps?

**Comment by ZLLentz** [comment]
I think changing the default to be more similar to the original behavior is wise. Note that it isn't quite exactly the same- previously we resolved the set without creating a thread if `delay == 0`. The tests still pass so I think this is :+1:

**Comment by tacaswell** [comment]
@callumforrester This needs a rebase.

**Comment by callumforrester** [comment]
@tacaswell Not sure why 3.9 hung, mind restarting?

**Comment by tacaswell** [comment]
Hmm, this is already restarted once.

**Pull request merged**